### ISSUE TITLE
x/net/http2: responseWriterState set to dirty incorrectly in high load performance test

### ISSUE THREAD

**Issue by crowfrog** [golang/go#26956]
### What version of Go are you using (`go version`)?
golang:1.10.0 linux

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
container: rhel
kubenet: 

### What did you do?
```
func (sc *serverConn) wroteFrame(res frameWriteResult) {
...
			sc.closeStream(st, errHandlerComplete)                  <=== _code1_
		}
...
	wr.replyToWriter(res.err)                                               <=== _code2_
...
}
```

**In some cases, the server goroutine will be switch out execution context between _code1_ and _code2_.
If the handler goroutine switch back to execution context in these cases, that will make writeDataFromHandler() (writeHeaders() also?) failed and set responseWriterState to dirty.**

```
func (sc *serverConn) writeDataFromHandler(stream *stream, data []byte, endStream bool) error {
...
	case <-stream.cw:
		// If both ch and stream.cw were ready (as might
		// happen on the final Write after an http.Handler
		// ends), prefer the write result. Otherwise this
		// might just be us successfully closing the stream.
		// The writeFrameAsync and serve goroutines guarantee
		// that the ch send will happen before the stream.cw
		// close.
		select {
		case err = <-ch:
			frameWriteDone = true
		default:
			return errStreamClosed
		}
	}
...
}
func (rws *responseWriterState) writeChunk(p []byte) (n int, err error) {
...
		if err := rws.conn.writeDataFromHandler(rws.stream, p, endStream); err != nil {
			rws.dirty = true
			return 0, err
		}
...
}
```

**This issue can be reproduced easily if add sc.logf() between _code1_ and _code2_ in lower load performace test.
PS: I think in wroteFrame(), we shouldn'd call "wr.replyToWriter(res.err)" when closing stream with error. It will cause writeDataFromHandler() exit with no error. But in fact this stream/responseWriterState is already on incorrect status.**

### What did you expect to see?
no connection reset in load test

### What did you see instead?
connecttion reset and traffic failed beofre new connection set up

**Comment by andybons** [comment]
@bradfitz @tombergan

**Comment by bradfitz** [comment]
Got a repro?

**Comment by fraenkel** [comment]
I have tried a variety of things with h2load as the client and a server that returns 100 - 16k bytes of data in the response. I have not been able to cause any connection reset issue regardless of debug on the server side to cause significant slow downs.

**Comment by crowfrog** [comment]
hi Fraenkel,

This issue can be reproduced easily if add sc.logf() between code1 and code2 in lower load performace test.
You need lots of streams, which with less data to transport, in one connection to reproduce it.
Such as, one connection, 3000 or more streams per second created, transported then closed.

B.R

**Comment by fraenkel** [comment]
@crowfrog I have added sc.logf() in multiple places between your code1 and code2 blocks. I still don't hit any error when driving with h2load.
It shows that I am achieving 7150 req/s.

I have also tried with and without debug to slow things down but that too doesn't cause any error.

**Comment by fraenkel** [comment]
@crowfrog Yes. There were over 10k streams created. I tried smallish and larger payloads but all constant sizes.
If you can capture http2debug, that might prove helpful to narrow down the issue since I cannot seem to do it locally on the latest code base.
